File: frontend/src/utils/errorHandler.ts
/**
 * Centralized Error Handler
 * Provides consistent, user-friendly error handling for SaaS CRM application
 */

import { toast } from 'react-hot-toast';

export interface ErrorHandlerOptions {
  showToast?: boolean;
  toastMessage?: string;
  logToConsole?: boolean;
  onError?: (error: any) => void;
}

export interface UserFriendlyError {
  title: string;
  message: string;
  suggestion?: string;
  status_code?: number;
  technical_detail?: string;
}

/**
 * Extract user-friendly error from API response
 */
const extractUserFriendlyError = (error: any): UserFriendlyError | null => {
  // Check if backend sent user-friendly error format
  if (error.response?.data?.error) {
    return error.response.data.error;
  }
  return null;
};

/**
 * Get fallback user-friendly message based on status code
 */
const getFallbackMessage = (statusCode: number, detail: string): UserFriendlyError => {
  switch (statusCode) {
    case 401:
      return {
        title: 'Session Expired',
        message: 'Your session has expired. Please log in again.',
        suggestion: 'Redirecting to login page...',
        status_code: 401
      };
    case 403:
      return {
        title: 'Access Denied',
        message: 'You don\'t have permission to perform this action.',
        suggestion: 'Contact your administrator if you need access to this feature.',
        status_code: 403
      };
    case 404:
      return {
        title: 'Not Found',
        message: 'The requested resource could not be found.',
        suggestion: 'The item may have been deleted or you may not have access to it.',
        status_code: 404
      };
    case 422:
      return {
        title: 'Validation Error',
        message: detail || 'Please check your input and try again.',
        suggestion: 'Make sure all required fields are filled correctly.',
        status_code: 422
      };
    case 500:
      return {
        title: 'Server Error',
        message: 'Something went wrong on our end.',
        suggestion: 'Please try again in a few moments or contact support.',
        status_code: 500
      };
    default:
      return {
        title: 'Error',
        message: detail || 'An unexpected error occurred.',
        suggestion: 'Please try again or contact support if the problem persists.',
        status_code: statusCode
      };
  }
};

/**
 * Handle API errors consistently with user-friendly messages
 */
export const handleApiError = (
  error: any,
  options: ErrorHandlerOptions = {}
): void => {
  const {
    showToast = true,
    toastMessage,
    logToConsole = false, // Changed default to false for production
    onError
  } = options;

  // Log to console only in development
  if (logToConsole || import.meta.env.DEV) {
    console.error('API Error:', error);
  }

  // Try to extract user-friendly error from backend
  const friendlyError = extractUserFriendlyError(error);
  
  if (friendlyError) {
    // Backend sent user-friendly error
    if (showToast) {
      const message = friendlyError.suggestion 
        ? `${friendlyError.title}: ${friendlyError.message}\n💡 ${friendlyError.suggestion}`
        : `${friendlyError.title}: ${friendlyError.message}`;
      toast.error(message, { duration: 5000 });
    }
  } else {
    // Fallback to extracting from standard error response
    const statusCode = error.response?.status || 500;
    const detail = error.response?.data?.detail || error.response?.data?.message || error.message;
    const fallbackError = getFallbackMessage(statusCode, detail);
    
    if (showToast) {
      const message = toastMessage || (
        fallbackError.suggestion 
          ? `${fallbackError.title}: ${fallbackError.message}\n💡 ${fallbackError.suggestion}`
          : `${fallbackError.title}: ${fallbackError.message}`
      );
      toast.error(message, { duration: 5000 });
    }
  }

  // Handle 401 - redirect to login
  if (error.response?.status === 401) {
    setTimeout(() => {
      localStorage.removeItem('token');
      window.location.href = '/login';
    }, 2000);
  }

  // Call custom error handler if provided
  if (onError) {
    onError(error);
  }
};

/**
 * Wrapper for async operations with error handling
 */
export const withErrorHandler = async <T>(
  operation: () => Promise<T>,
  options: ErrorHandlerOptions = {}
): Promise<T | null> => {
  try {
    return await operation();
  } catch (error) {
    handleApiError(error, options);
    return null;
  }
};

/**
 * Handle form validation errors
 */
export const handleValidationError = (
  field: string,
  message: string
): void => {
  toast.error(`${field}: ${message}`);
};

File: backend/app/middleware/error_handler.py
"""
User-friendly error handler middleware for permission and access control errors
Converts technical errors into user-friendly messages for a SaaS CRM application
"""

from fastapi import Request, status
from fastapi.responses import JSONResponse
from fastapi.exceptions import HTTPException
import logging

logger = logging.getLogger(__name__)


class UserFriendlyError:
    """User-friendly error messages for common permission scenarios"""
    
    # Permission Errors
    PERMISSION_DENIED = {
        "title": "Access Denied",
        "message": "You don't have permission to perform this action.",
        "suggestion": "Please contact your administrator if you need access to this feature."
    }
    
    VIEW_COMPANY_DATA = {
        "title": "Company Data Access Required",
        "message": "You can only view data within your team or assigned to you.",
        "suggestion": "Contact your Company Admin to request broader access permissions."
    }
    
    VIEW_TEAM_DATA = {
        "title": "Team Access Required",
        "message": "You can only view your own data and tasks.",
        "suggestion": "Ask your Sales Manager to assign you to a team for team-level access."
    }
    
    MANAGE_USERS = {
        "title": "User Management Restricted",
        "message": "Only administrators can add or manage users.",
        "suggestion": "Contact your Company Admin to add new team members."
    }
    
    MANAGE_BILLING = {
        "title": "Billing Access Restricted",
        "message": "Only Company Admins can manage billing and subscriptions.",
        "suggestion": "Contact your Company Admin for billing-related requests."
    }
    
    MANAGE_COMPANY = {
        "title": "Company Settings Restricted",
        "message": "Only Company Admins can modify company settings.",
        "suggestion": "Contact your administrator to update company information."
    }
    
    ASSIGN_LEADS = {
        "title": "Lead Assignment Restricted",
        "message": "You don't have permission to assign leads to other users.",
        "suggestion": "Contact your Sales Manager to reassign leads or deals."
    }
    
    EXPORT_DATA = {
        "title": "Data Export Restricted",
        "message": "You don't have permission to export company data.",
        "suggestion": "Contact your Company Admin to request a data export."
    }
    
    MANAGE_INTEGRATIONS = {
        "title": "Integration Management Restricted",
        "message": "Only administrators can configure integrations.",
        "suggestion": "Contact your Company Admin to set up or modify integrations."
    }
    
    CUSTOMIZE_CRM = {
        "title": "CRM Customization Restricted",
        "message": "Only administrators can customize CRM fields and settings.",
        "suggestion": "Contact your Company Admin to request custom fields or pipeline changes."
    }
    
    # Team Errors
    NO_TEAM_ASSIGNED = {
        "title": "No Team Assigned",
        "message": "You are not currently assigned to any team.",
        "suggestion": "Contact your administrator to be added to a team."
    }
    
    TEAM_ACCESS_DENIED = {
        "title": "Team Access Denied",
        "message": "You can only access your own team's data.",
        "suggestion": "Contact your Sales Manager if you need access to other teams."
    }
    
    # Resource Errors
    RESOURCE_NOT_FOUND = {
        "title": "Not Found",
        "message": "The requested resource could not be found.",
        "suggestion": "The item may have been deleted or you may not have access to it."
    }
    
    CONTACT_NOT_FOUND = {
        "title": "Contact Not Found",
        "message": "This contact doesn't exist or you don't have access to it.",
        "suggestion": "Check if the contact was deleted or contact your team lead."
    }
    
    DEAL_NOT_FOUND = {
        "title": "Deal Not Found",
        "message": "This deal doesn't exist or you don't have access to it.",
        "suggestion": "Check if the deal was deleted or contact your team lead."
    }
    
    # Company Errors
    COMPANY_NOT_FOUND = {
        "title": "Company Not Found",
        "message": "The requested company could not be found.",
        "suggestion": "Please check the company ID or contact support."
    }
    
    SUBSCRIPTION_EXPIRED = {
        "title": "Subscription Expired",
        "message": "Your company's subscription has expired.",
        "suggestion": "Contact your Company Admin to renew the subscription."
    }
    
    TRIAL_EXPIRED = {
        "title": "Trial Period Ended",
        "message": "Your free trial has ended.",
        "suggestion": "Upgrade to a paid plan to continue using all features."
    }
    
    # Authentication Errors
    INVALID_TOKEN = {
        "title": "Session Expired",
        "message": "Your session has expired. Please log in again.",
        "suggestion": "Click here to return to the login page."
    }
    
    ACCOUNT_DISABLED = {
        "title": "Account Disabled",
        "message": "Your account has been disabled.",
        "suggestion": "Contact your administrator or support for assistance."
    }


def get_user_friendly_error(error_detail: str, status_code: int) -> dict:
    """
    Convert technical error messages to user-friendly format
    """
    error_lower = error_detail.lower()
    
    # Permission-related errors
    if "permission" in error_lower or "forbidden" in error_lower:
        if "company data" in error_lower or "company" in error_lower:
            return UserFriendlyError.VIEW_COMPANY_DATA
        elif "team" in error_lower:
            if "access denied to this team" in error_lower:
                return UserFriendlyError.TEAM_ACCESS_DENIED
            return UserFriendlyError.VIEW_TEAM_DATA
        elif "user" in error_lower or "member" in error_lower:
            return UserFriendlyError.MANAGE_USERS
        elif "billing" in error_lower or "subscription" in error_lower:
            return UserFriendlyError.MANAGE_BILLING
        elif "assign" in error_lower or "lead" in error_lower:
            return UserFriendlyError.ASSIGN_LEADS
        elif "export" in error_lower or "import" in error_lower:
            return UserFriendlyError.EXPORT_DATA
        elif "integration" in error_lower:
            return UserFriendlyError.MANAGE_INTEGRATIONS
        elif "customize" in error_lower or "field" in error_lower:
            return UserFriendlyError.CUSTOMIZE_CRM
        else:
            return UserFriendlyError.PERMISSION_DENIED
    
    # Team-related errors
    if "no team" in error_lower or "not assigned to" in error_lower:
        return UserFriendlyError.NO_TEAM_ASSIGNED
    
    # Resource not found errors
    if status_code == 404:
        if "contact" in error_lower:
            return UserFriendlyError.CONTACT_NOT_FOUND
        elif "deal" in error_lower:
            return UserFriendlyError.DEAL_NOT_FOUND
        elif "company" in error_lower:
            return UserFriendlyError.COMPANY_NOT_FOUND
        else:
            return UserFriendlyError.RESOURCE_NOT_FOUND
    
    # Subscription errors
    if "subscription" in error_lower and "expired" in error_lower:
        return UserFriendlyError.SUBSCRIPTION_EXPIRED
    
    if "trial" in error_lower and "expired" in error_lower:
        return UserFriendlyError.TRIAL_EXPIRED
    
    # Authentication errors
    if status_code == 401:
        if "disabled" in error_lower:
            return UserFriendlyError.ACCOUNT_DISABLED
        return UserFriendlyError.INVALID_TOKEN
    
    # Default error
    return {
        "title": "Error",
        "message": error_detail,
        "suggestion": "Please try again or contact support if the problem persists."
    }


async def error_handler_middleware(request: Request, call_next):
    """
    Middleware to catch errors and convert them to user-friendly messages
    """
    try:
        response = await call_next(request)
        return response
    except HTTPException as exc:
        # Convert to user-friendly error
        friendly_error = get_user_friendly_error(exc.detail, exc.status_code)
        
        # Log the original error for debugging
        logger.warning(
            f"HTTP {exc.status_code} - {exc.detail} - "
            f"Path: {request.url.path} - User: {request.state.user if hasattr(request.state, 'user') else 'Unknown'}"
        )
        
        return JSONResponse(
            status_code=exc.status_code,
            content={
                "error": {
                    "title": friendly_error["title"],
                    "message": friendly_error["message"],
                    "suggestion": friendly_error["suggestion"],
                    "status_code": exc.status_code,
                    "technical_detail": exc.detail if request.state.get("is_admin", False) else None
                }
            }
        )
    except Exception as exc:
        # Log unexpected errors
        logger.error(f"Unexpected error: {str(exc)}", exc_info=True)
        
        return JSONResponse(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            content={
                "error": {
                    "title": "Something Went Wrong",
                    "message": "An unexpected error occurred. Our team has been notified.",
                    "suggestion": "Please try again in a few moments or contact support if the issue persists.",
                    "status_code": 500
                }
            }
        )
